instrumentation/apache-httpclient/apache-httpclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/async/ApacheHttpClientRequest.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.apachehttpclient.async;
+
+import static java.util.logging.Level.FINE;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+import org.apache.hc.core5.http.EntityDetails;
+import org.apache.hc.core5.http.Header;
+import org.apache.hc.core5.http.HttpRequest;
+
+public final class ApacheHttpClientRequest {
+
+  private static final Logger logger = Logger.getLogger(ApacheHttpClientRequest.class.getName());
+
+  @Nullable private final URI uri;
+
+  private final HttpRequest delegate;
+  private final EntityDetails entityDetails;
+
+  public ApacheHttpClientRequest(HttpRequest httpRequest, EntityDetails entityDetails) {
+    uri = getUri(httpRequest);
+    this.entityDetails = entityDetails;
+    this.delegate = httpRequest;
+  }
+
+  public List<String> getHeader(String name) {
+    return headersToList(delegate.getHeaders(name));
+  }
+
+  // minimize memory overhead by not using streams
+  static List<String> headersToList(Header[] headers) {
```

Comment:
same with this method, if exactly the same as method in classic

Reply:
The suggestions are good, but I guess it will involve more refactoring in both parts of the client support.
Like when I started this MR, I thought of just using HttpRequest only which is http client specific, but then I got a use-case of keeping the extra class AsyncHttpClientRequest which was to store the entity.

Both the async and the http client are kind of using different strategies. In async we are decorating, in classic we are replacing the request instance. So better way would be to have something in common and another way could be to have something like Request / Response adapters which operates on HttpRequest / HttpResponse. I see other methods which repeat like getFlavor